r/tests/testthat/test-dplyr-funcs-datetime.R

```diff
@@ -2309,3 +2308,534 @@ test_that("build_formats() and build_format_from_order()", {
+    call_binding("round_date", Expression$scalar(Sys.time()), "61 seconds"),
+    "Rounding with second > 60 is not supported"
+  )
+  expect_error(
+    call_binding("round_date", Expression$scalar(Sys.time()), "61 minutes"),
+    "Rounding with minute > 60 is not supported"
+  )
+  expect_error(
+    call_binding("round_date", Expression$scalar(Sys.time()), "25 hours"),
+    "Rounding with hour > 24 is not supported"
+  )
+  expect_error(
+    call_binding("round_date", Expression$scalar(Sys.Date()), "25 hours"),
+    "Rounding with hour > 24 is not supported"
+  )
+
+})
+
+check_timezone_rounding <- function(data, unit) {
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        utc_floored = floor_date(utc_time, unit = unit),
+        utc_rounded = round_date(utc_time, unit = unit),
+        utc_ceiling = ceiling_date(utc_time, unit = unit),
+        syd_floored = floor_date(syd_time, unit = unit),
+        syd_rounded = round_date(syd_time, unit = unit),
+        syd_ceiling = ceiling_date(syd_time, unit = unit),
+        adl_floored = floor_date(adl_time, unit = unit),
+        adl_rounded = round_date(adl_time, unit = unit),
+        adl_ceiling = ceiling_date(adl_time, unit = unit),
+        mar_floored = floor_date(mar_time, unit = unit),
+        mar_rounded = round_date(mar_time, unit = unit),
+        mar_ceiling = ceiling_date(mar_time, unit = unit),
+        kat_floored = floor_date(kat_time, unit = unit),
+        kat_rounded = round_date(kat_time, unit = unit),
+        kat_ceiling = ceiling_date(kat_time, unit = unit)
+      ) %>%
+      collect(),
+    data
+  )
+}
+
+test_that("timestamp rounding takes place in local time", {
+
+  # lubridate does not always return the correct results
+  # for some of our test cases in tz_times, esp on windows
+  skip_on_os("windows")
+
+  tz_times %>% check_timezone_rounding(".001 second")
+  tz_times %>% check_timezone_rounding("second")
+  tz_times %>% check_timezone_rounding("minute")
+  tz_times %>% check_timezone_rounding("hour")
+
+})
+
+test_that("timestamp rounding takes place in local time", {
+
+  # a less strict test than tz_times
+  year_of_dates_tz <- tibble::tibble(
+    utc_time = as.POSIXct(month_boundaries, tz = "UTC"),
+    syd_time = as.POSIXct(month_boundaries, tz = "Australia/Sydney"),
+    adl_time = as.POSIXct(month_boundaries, tz = "Australia/Adelaide"),
+    mar_time = as.POSIXct(month_boundaries, tz = "Pacific/Marquesas"),
+    kat_time = as.POSIXct(month_boundaries, tz = "Asia/Kathmandu")
+  )
+
+  year_of_dates_tz %>% check_timezone_rounding(".001 second")
+  year_of_dates_tz %>% check_timezone_rounding("second")
+  year_of_dates_tz %>% check_timezone_rounding("minute")
+  year_of_dates_tz %>% check_timezone_rounding("hour")
+  year_of_dates_tz %>% check_timezone_rounding("day")
+  year_of_dates_tz %>% check_timezone_rounding("month")
+  year_of_dates_tz %>% check_timezone_rounding("quarter")
+  year_of_dates_tz %>% check_timezone_rounding("year")
+
+})
```

Comment:
I'm afraid to ask, but what about `week`?

Reply:
oops I forgot to list that case. added now!